### PR TITLE
Feature/supertest fixes #22

### LINKED ISSUE #22: Supertest to test express integration
Is anyone against using this instead of the bash script in use currently?

If not, expect a PR.

In fact expect one anyway, no way I want to contribute unit tests if I have to copy-paste bash

### PR DESCRIPTION
**Pull request by faceleg**
This fixes #22 

[Trello Card](https://trello.com/c/0qhPLIXB/66-seneca-web-feature-supertest-fixes-22)

### CODE REVIEW

**Comment by faceleg** [comment]
Ready for y'all to tear apart

**Comment by faceleg** [comment]
Ideally (IMHO) `npm run lint` would be run as part of the build, but baby steps

**Comment by mcdonnelldean** [comment]
@faceleg This looks nice, I'll further review in the morning.

One thing I would like to see is a move towards the org standard that is happening; best seen in the level store https://github.com/senecajs/seneca-level-store. I'd rather not bring gulp in right now as we are trying to keep the testing (with lab) and the linting (eslint) super simple. We are following Standard's rules for the repo updates we are doing.

Regardless, the supertest stuff looks pretty epic :D I'll take a look in detail first thing in the morning, but good work :+1:

**Comment by faceleg** [comment]
I'll remove the gulp bits, no problem

**Comment by mcdonnelldean** [comment]
@faceleg Awesome :D I'm a gulp user myself but have to keep it super simple for org repos for new contributors. If you check the level store you will see it's eslint too, it will make yours smaller but it's more or less the same rules. As I said though, super work! This is why I love Open Source :D

**Comment by faceleg** [comment]
IMHO the gulpfile makes it even easier for new contributors, as the deps are installed with the npm test pretest command, husky automatically sets up the git hooks, and the linting / tests are run ON COMMIT, so there is less friction for new devs as they don't actually need to do anything - their computer tells them if they've done it all right, in terms of tests passing / linter linting.

But I'll be happy conforming to whatever standards or opinions you "The Project" requires, and won't be so presumptive again :)

**Comment by faceleg** [comment]
I'll make another feature branch based on this to work on #21

**Comment by mcdonnelldean** [comment]
@faceleg No no!!

Never stop being presumptive! Sometimes org rules take precedent and that's ok but other times, like with supertest, we never even considered it. We love having people contribute, especially improvements. We love when people reach out and take ownership like this; we positively encourage it!

**Comment by mcdonnelldean** [comment]
In addition, if you do have any opinions or thoughts on org wide issues we would love to hear them, stuff like refining building and testing is welcome; we even have a special place set up for it here,

https://github.com/senecajs/contrib

If you want to add to the issues there we can get the ball rolling and a discussion going. As I said sometimes ideas won't fit.... but other times they do :smile:

**Comment by faceleg** [comment]
Opened https://github.com/senecajs/contrib/issues/34

**Comment by mcdonnelldean** [comment]
@faceleg Awesome :dancer:

**Comment by faceleg** [comment]
Related: senecajs/contrib#35

File: package.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
   "license": "MIT",
   "main": "web.js",
   "scripts": {
+    "pretest": "npm install",
```
Comment (mcdonnelldean) [inline package.json]: Awesome! @geek we should do this on all our package.json's

Comment (faceleg) [inline package.json]: You can "pre" just about anything, very useful.

e.g.

```JSON
  "preprotractor": "start the engines",
  "protractor": "some command"
```

**Comment by faceleg** [comment]
Soooo the tests fail in node 0.10 because "Promise" is not defined.

Options:

1. Remove use of promises 
2. Use `bluebird` in dev 
3. Remove node 0.10 from .travis.yml

@mcdonnelldean @rjrodger

**Comment by faceleg** [comment]
Is option 2 going to work for you all?

@rjrodger @geek @mcdonnelldean

**Comment by geek** [comment]
Does it need to use promises?  If it does, then bluebird sounds like a plan.

**Comment by faceleg** [comment]
I can't argue that it *must* use promises. Given a choice between a promise and a callback, I will choose promise everytime.

**Comment by mcdonnelldean** [comment]
@geek @faceleg Was it using promises before this? How was it doing it previous? I've heard good things about bluebird so if we need it I'm :+1: I just want to see what was it before? Sorry, ramping up.

Is it just because of node .10 ?

**Comment by faceleg** [comment]
The relevant file (test/example.js) didn't use any form of async. 

I made the current change so the tests can get the `app` variable, which `supertest` needs in order to be able to test an express app.

If one runs it naked, it'll start listening on port 3000 as it did before.

Prior to my change, the example.js file did not export anything, as any tests were done using bash.

I've used bluebird for a good long time, it's great. I originally intended just to use native promises for this simple implementation, but as Promise doesn't exist in 0.10, I had to drop bluebird in there.

The alternative is to refactor this portion to use callbacks instead of promises, but I humbly submit to you that promises are so much superior to callbacks in terms of readability composability and maintainability that we should be promoting their use wherever possible.

**Comment by mcdonnelldean** [comment]
Cheers! for the response. Let me give this all a look over tomorrow. 

Thanks a mill for the work so far :D On light review it all looks awesome :D

**Comment by faceleg** [comment]
No problem, supremely happy to have the chance to contribute

**Comment by mcdonnelldean** [comment]
@faceleg First off, Awesome work! This is a really nice PR.

The only thing it needs is a good lint, there are some semi-colons and additional spaces in there. If you can get these knocked out and hook up linting to be forced on `npm test` that would be awesome, we can get it merged for ya :D

**Comment by faceleg** [comment]
When I run `standard` against the repo I get hundreds of lines of output, including for files I haven't touched.

I'll fix it, but it'll take time.

**Comment by faceleg** [comment]
Witness: https://gist.github.com/faceleg/460b0e1d4146cb38876c

**Comment by mcdonnelldean** [comment]
Only do the ones you touched, for the others, ignore them with a .eslintignore you don't need to do the whole rep :D

**Comment by faceleg** [comment]
:P

**Comment by faceleg** [comment]
Shall I close this?

**Comment by mcdonnelldean** [comment]
@faceleg No no, can you make it remergable?

**Comment by geek** [comment]
@faceleg can you update to be mergable?

**Comment by faceleg** [comment]
Doing so now

**Comment by faceleg** [comment]
@geek ready to go

**Comment by faceleg** [comment]
Thanks @geek!

**Comment by geek** [comment]
@faceleg should be published tomorrow, thank you

**Pull request merged**